examples/char-rnn.cpp

```diff
@@ -0,0 +1,260 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "glow/ExecutionEngine/ExecutionEngine.h"
+#include "glow/Graph/Graph.h"
+#include "glow/IR/IR.h"
+#include "glow/Support/Support.h"
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Format.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Timer.h"
+
+#include <string>
+
+//----------------------------------------------------------------------------//
+// This is a small program that's based on Andrej's char-rnn generator. This is
+// a small RNN-based neural network that's used to generate random text after
+// analyzing some other text. The network is described here:
+// http://karpathy.github.io/2015/05/21/rnn-effectiveness/
+//----------------------------------------------------------------------------//
+
+using namespace glow;
+using llvm::format;
+
+namespace {
+llvm::cl::OptionCategory category("char-rnn Options");
+static llvm::cl::opt<std::string> inputFilename(llvm::cl::desc("input file"),
+                                                llvm::cl::init("-"),
+                                                llvm::cl::Positional,
+                                                llvm::cl::cat(category));
+
+llvm::cl::opt<BackendKind> executionBackend(
+    llvm::cl::desc("Backend to use:"),
+    llvm::cl::values(clEnumValN(BackendKind::Interpreter, "interpreter",
+                                "Use interpreter (default option)"),
+                     clEnumValN(BackendKind::CPU, "cpu", "Use CPU"),
+                     clEnumValN(BackendKind::OpenCL, "opencl", "Use OpenCL")),
+    llvm::cl::init(BackendKind::Interpreter), llvm::cl::cat(category));
+
+llvm::cl::opt<unsigned> numEpochs("epochs",
+                                  llvm::cl::desc("Process the input N times."),
+                                  llvm::cl::init(4), llvm::cl::value_desc("N"),
+                                  llvm::cl::cat(category));
+llvm::cl::opt<unsigned>
+    generateChars("chars", llvm::cl::desc("Generate this number of chars."),
+                  llvm::cl::init(1000), llvm::cl::value_desc("N"),
+                  llvm::cl::cat(category));
+
+} // namespace
+
+PseudoRNG RNG;
+
+/// Clip the value \p c to the range 0..127, which is standard ascii.
+static size_t clipASCII(char c) {
+  size_t c1 = c;
+  if (c1 > 127) {
+    c1 = 127;
+  }
+  return c1;
+}
+
+/// Load text into \p inputText that has the format  [B, S, 128], where B is
+/// the batch size, S is the length of the sentence, and 128 is the one-hot
```

Comment:
"the one-hot representation of the text" can you explain this a bit.

Reply:
also, maybe make 128 a const and use it everywhere.

Reply:
I am adding a link to wikipedia's definition of one-hot. It is commonly used in neural networks and electronics, but it can't hurt to add a link.